### ISSUE TITLE
Allow tests to set the tracing log level

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#760]
We should add a function to zebra-test that allows tests to set the tracing log level, or disable logs entirely.

Then we can modify the tests as follows:
- [x] tests marked "should_panic" should disable all output (yaahc/color-eyre#42, #759)
- [ ] tests that log errors should disable non-panic output (set the log level to 'off')
- [ ] tests that log warnings should set the log level to 'error'

TODO:
- [x] Set the default test log level to 'warn'
- [x] Implement a crate-level workaround
- [ ] Make the tracing log level test-specific, rather than having it apply to the whole crate

**Comment by yaahc** [comment]
I don't think we should disable output on tests marked `should_panic`, also it wouldn't fix the issue. It was a bug in the behaviour of `color-eyre` that I've fixed in https://github.com/yaahc/color-eyre/pull/42. It was actually a really interesting issue so you should check out the description in PR that fixes it. also if you don't mind I'd love your review on the code.

**Comment by teor2345** [comment]
Thanks, I've ticked off the panic case, now I just have to work out how to handle the warning case in some of the zebra-consensus tests.

**Comment by yaahc** [comment]
it likely comes from output from spawned threads. the output capture is controlled by a thread local that is set once before the start of each test. however thread locals on new threads have to be initialized to a default value, there's no way to set it based on the thread that spawned it, so new threads end up with `LOCAL_STD*` pointing back at the real io handles.

I'm not sure what options there are for making tracing output via the original thread. I think there is a form of subscriber that runs in its own thread, or at least plans for one. we'd want to use that when we run our tests and ensure that the thread doing the logging is the original thread for our test, and run the rest of our test in spawned threads, for multithreaded tests.

cc @hawkw

**Comment by teor2345** [comment]
I'm talking about tracing INFO and WARN level logs, for example:

    Jul 24 03:32:34.537  WARN block_higher_than_max_checkpoint_fail::comment{text=Set up the future for block 415000}: error="block is higher than the maximum checkpoint"
    Jul 24 03:32:34.541  INFO checkpoint_drop_cancel::comment{height=BlockHeight(0) text=Set up the future for block}: Queued checkpoint block height=BlockHeight(0) hash=BlockHeaderHash("08ce3d9731b000c08338455c8a4a6bd05da16e26b11daa1b917184ece80f0400") is_checkpoint=true

https://github.com/ZcashFoundation/zebra/pull/748/checks?check_run_id=905260423#step:5:909

So if my understanding is correct, we should be able to add those spans to the color-eyre filter, and they will disappear from the output:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-test/src/lib.rs#L23

If that doesn't work, we could set the default tracing filter to exclude those logs.

**Comment by yaahc** [comment]
The `color-eyre` filter currently only applies to frames in backtraces printed in the report / panic handlers so you're gonna need to use a tracing filter here.

**Comment by yaahc** [comment]
@hawkw has suggested using the `MakeWriter` api to essentially reimplement test capturing. We can use a writer that buffers the log output the same way the tests would and use a drop guard to print the logs, that drop guard would be owned by the main thread and would always be the last thing dropped and would always drop in the correct thread so it would end up being captured by the built in capturing. This should be relatively easy to implement ontop of our existing `zebra_test::init()` logic.

**Comment by teor2345** [comment]
Sounds good, but do we lose the ability to see incremental test output?

That matters if the test fails in a way that affects the output or the process. For example, a double-panic which crashes the whole process.

If we could lose logs like this, let's check an environmental variable before enabling filtering.

**Comment by yaahc** [comment]
> Sounds good, but do we lose the ability to see incremental test output?
> 
> That matters if the test fails in a way that affects the output or the process. For example, a double-panic which crashes the whole process.
> 
> If we could lose logs like this, let's check an environmental variable before enabling filtering.

I think it will lose less incremental test output than disabling tracing all together would. Actually, now that I think of it I think we test with `panic=abort` so we wouldn't end up dropping the tracing logger anyways if we panicked.

**Comment by hawkw** [comment]
> > 
> 
> I think it will lose less incremental test output than disabling tracing all together would. Actually, now that I think of it I think we test with `panic=abort` so we wouldn't end up dropping the tracing logger anyways if we panicked.

Oh, in that case, this approach isn't really going to work at all.

**Comment by teor2345** [comment]
Could we drop the tracing logger in the panic handler?

Or does that cause threading issues?

**Comment by teor2345** [comment]
Or could we use a writer that line-buffers test output?

**Comment by yaahc** [comment]
@teor2345 we probably insert a clone of the guard into some global variable that the panic handler then accesses and uses to flush the output before aborting.

**Comment by teor2345** [comment]
I've just opened PR #761, which sets two filters if `$RUST_LOG` is not set:
* `warn`
* `zebra_consensus=error`

I tried adding specific spans to the zebra_consensus filter, but they didn't work. I don't know if I'm using the wrong tracing syntax, or if it's not possible to filter using spans.

In any case, this PR fixes the two remaining issues I had. (I also lose any other warning logs in `zebra_consensus`, but I can get them back using `RUST_LOG=warn`.)
